zio-config/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -1,13 +1,52 @@
 package zio.config
 
-import zio.IO
+import zio.system.System
+import zio.{ IO, ZIO }
 
-trait ConfigSource {
-  val configService: ConfigSource.Service
+trait ConfigSource[K, V] {
+  val configSourceService: ConfigSource.Service[K, V]
 }
 
 object ConfigSource {
-  trait Service {
-    def getConfigValue(path: String): IO[ReadError, String]
+  trait Service[K, V] {
+    def getConfigValue(path: K): IO[ReadError[K, V], PropertyTree[K, V]]
```

Comment:
This is the "non-flattened" representation which may be more difficult to support for, e.g. system properties, environment variables, etc. 

The "flattened" representation is:

```scala
def getConfigValue(path: List[K]): IO[ReadError[K, V], V]
```

This representation permits the possibility that intermediate nodes cannot be read; only terminal nodes can be read. Which is true for completely flat sources with emulated nesting (e.g. `foo.bar.baz = 123`).